### PR TITLE
[4.0] Implement the getName function in the ConsoleApplication class

### PR DESCRIPTION
**Pull request by twister65**
Pull Request for Issue https://github.com/joomla/joomla-cms/issues/23845.

### Summary of Changes
Use the getName function instead of getClientId function in the system/debug plugin.
Add the getName function in the ConsoleApplication class.
See the @SniperSister comment:
https://github.com/joomla/joomla-cms/pull/23859#issuecomment-492910869

### Testing Instructions
Enable Debug System.
And try:
`php cli/joomla.php list`

### CODE REVIEW

File: libraries/src/Application/ConsoleApplication.php [hunk 1]
```diff
@@ -81,6 +89,9 @@ public function __construct(
 		$this->setName('Joomla!');
 		$this->setVersion(JVERSION);
 
+		// Register the client ID as administrator
+		$this->clientId = 1;
```
Comment (mbabker) [inline libraries/src/Application/ConsoleApplication.php]: 👎

Each application should have a unique ID.  Spoofing IDs should not be encouraged.

Note, the client also needs to be registered in [`ApplicationHelper::getClientInfo()`](https://github.com/joomla/joomla-cms/blob/ffe41c4e702818528c625e5194c961a088dc8a74/libraries/src/Application/ApplicationHelper.php#L123).

Comment (twister65) [reply libraries/src/Application/ConsoleApplication.php]: Should we use client Id '4' and client name 'cli' ?

Comment (mbabker) [reply libraries/src/Application/ConsoleApplication.php]: That'd make the most sense.

Comment (twister65) [reply libraries/src/Application/ConsoleApplication.php]: On the other hand, the cache path is './administrator/cache/plg_system_debug_1' when we enable debug system.
1 <=> administrator

File: libraries/src/Application/CMSApplicationInterface.php [hunk 2]
```diff
@@ -155,6 +155,15 @@ public function isCli();
 	 */
 	public function getIdentity();
 
+	/**
+	 * Gets the client id of the current running application.
+	 *
+	 * @return  integer  A client identifier.
+	 *
+	 * @since   __DEPLOY_VERSION_
```
Comment (Quy) [inline libraries/src/Application/CMSApplicationInterface.php]:
```suggestion
	 * @since   __DEPLOY_VERSION__
```
Repeat in the other files.

File: libraries/src/Application/ConsoleApplication.php (hunk 1 above)
Comment (mbabker) [reply libraries/src/Application/ConsoleApplication.php]: As of 4.0 all system cache is in the `administrator/cache` directory.

Comment (mbabker) [reply libraries/src/Application/ConsoleApplication.php]: If there is something in the CLI integration that uses the info the debug plugin caches, then that part of the integration should do what it needs to to resolve the right names.  The CLI app should not reuse an identifier for convenience.  It's part of the same reasoning `isSite()` and `isAdmin()` are gone.

**Comment by SharkyKZ** [comment]
IMO, would be better to use application name. ID is so meaningless.

**Comment by richard67** [comment]
@twister65 When doing a `php cli/joomla.php list`, I get a PHP fatal error in any case, with and without your PR applied, and with and without the debug system on:

> PHP Fatal error:  Declaration of Joomla\CMS\Application\ConsoleApplication::doExecute() must be compatible with Joomla\Console\Application::doExecute(): int in /home/richard/lamp/public_html/joomla-cms-4.0-dev/libraries/src/Application/ConsoleApplication.php on line 32

Am I making some mistake?

**Comment by mbabker** [comment]
https://github.com/joomla/joomla-cms/issues/25374

TL;DR someone needs to update the console application class to add missing typehints added to new API since the PHP minimum was raised from 7.0 to 7.2.

**Comment by richard67** [comment]
Thanks Michael.

**Comment by richard67** [comment]
~~@mbabker and/or @wilsonge Am I wrong or does the composer.json of J4 point to a 2.0-dev branch of the framework's console package (joomla/console), and that branch does not exist, so the console package is never updated with a `composer update joomla/console` in J4?~~

Sorry, was wrong.

**Comment by twister65** [comment]
@richard67 , to test you can replace this line:
https://github.com/twister65/joomla-cms/blob/bd271abc484a5ac25a5d1d365d20440d9dafc3e9/libraries/src/Application/ConsoleApplication.php#L126
by:
`protected function doExecute(): int`

waiting for a new PR...

**Comment by richard67** [comment]
@mbabker Hmm, composer update for the console package does not do anything. Maybe my previous comment was not so wrong?

**Comment by richard67** [comment]
@twister65 To fix that 1 line leads to the next error, so it needs to pull the latest changes from the framework for the console package. But a `composer update` does not do anything, so I think there is something wrong, see my comments above. If that will be done, I am sure also #25374 will be fixed.

**Comment by mbabker** [comment]
If a `composer update` is not doing anything then that means the CMS already has the latest code.  The problem is not needing another `composer update`, the problem is the CMS subclass of `Joomla\Console\Application` needs updates to match changes from the parent repo.

**Comment by richard67** [comment]
I see, thanks again Michael.
To all: Is there already anybody working on that?

File: libraries/src/Application/ConsoleApplication.php [hunk 3]
```diff
@@ -81,6 +89,9 @@ public function __construct(
 		$this->setName('Joomla!');
 		$this->setVersion(JVERSION);
 
+		// Register the client ID as CLI
+		$this->clientId = ApplicationHelper::getClientInfo('cli', true)->id;
```
Comment (SharkyKZ) [inline libraries/src/Application/ConsoleApplication.php]: This seems unnecessary. What I meant by previous comment, it would be better to add `getName()` method to the interface and replace `getClientId()` call with it in Debug plugin. It's already present in `CMSApplication`.

**Comment by richard67** [comment]
PR for synching the function signatures in the subclass of the console application to the changes in the framework package as explained by Michael is PR #25391 . Thanks to @HLeithner for that.

@twister65 That PR is requirement for this PR here to be tested. Please test it.

**Comment by richard67** [comment]
@twister65 P.S.: When that other PR will be merged, you will have to rebase this here and solve merge conflicts.

File: libraries/src/Application/ConsoleApplication.php (hunk 3 above)
Comment (wilsonge) [reply libraries/src/Application/ConsoleApplication.php]: I agree with this statement

Comment (twister65) [reply libraries/src/Application/ConsoleApplication.php]: Done. Please, see the last commit https://github.com/joomla/joomla-cms/pull/25376/commits/262e350d9952a85973d3ac269dc4c0af4125d86c.

**Comment by richard67** [comment]
I have tested this item :white_check_mark: successfully on 262e350d9952a85973d3ac269dc4c0af4125d86c<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/25376">issues.joomla.org/tracker/joomla-cms/25376</a>.</sub>

**Comment by wilsonge** [comment]
Thanks!
